File: solutions/22 - Generate Parentheses.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        // Time Complexity: O(4^N/√N) because the approximation for the Nth Catalan number is
        //                  O(4^N/N^(3/2)) and string concatenation is O(N) so overall it is
        //                  O(4^N/√N).
        // Space Complexity: O(N) for recursion stack.
        vector<string> ans;
        recurse(ans, 0, n, "");
        return ans;
    }
    void recurse(vector<string>& ans, int left, int totalLeft, string curr)
    {
        if (totalLeft == 0)
        {
            ans.push_back(curr);
            return;
        }
        if (left == 0)
        {
            curr += "(";
            recurse(ans, left + 1, totalLeft, curr);
        }
        else if (left == totalLeft)
        {
            curr += ")";
            recurse(ans, left - 1, totalLeft - 1, curr);
        }
        else
        {
            recurse(ans, left - 1, totalLeft - 1, curr + ")");
            recurse(ans, left + 1, totalLeft, curr + "(");
        }
        return;
    }
};

int main() {
    int n;
    cout << "Enter n: ";
    cin >> n;

    Solution sol;
    vector<string> result = sol.generateParenthesis(n);

    cout << "All valid parentheses combinations:\n";
    for (const string& s : result) {
        cout << s << "\n";
    }
    cout << "Total combinations: " << result.size() << endl;

    return 0;
}

File: solutions/22 - Generate Parentheses.java
import java.util.*;
class Solution {
    public List<String> generateParenthesis(int n) {
        // Time Complexity: O(Cn * n) since the Catalan number is the amount of valid parentheses.
        //                  For each string we have to build it which is O(N).
        // Space Complexity: O(Cn * n) for Cn strings of length 2n.
        List<String> answer = new ArrayList<>();
        dfs(n, answer, 0, 0, "");
        return answer;
    }

    void dfs(int n, List<String> answer, int open, int end, String curr) {
        if (open == n && end == n) {
            answer.add(curr);
            return;
        }
        if (open < n) {
            dfs(n, answer, open + 1, end, curr + "(");
        }
        if (end < n && open > end) {
            dfs(n, answer, open, end + 1, curr + ")");
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.println("Enter n: ");
        int n = Integer.parseInt(sc.nextLine().trim());

        Solution sol = new Solution();
        List<String> result = sol.generateParenthesis(n);

        System.out.println("Generated parentheses: ");
        System.out.println(result);

        sc.close();
    }
}